src/ring_theory/henselian.lean

```diff
@@ -0,0 +1,247 @@
+  ∃ a : R, f.is_root a ∧ (a - a₀ ∈ maximal_ideal R))
+
+@[priority 100] -- see Note [lower instance priority]
+instance field.henselian (K : Type*) [field K] : henselian_local_ring K :=
+{ is_henselian := λ f hf a₀ h₁ h₂,
+  begin
+    refine ⟨a₀, _, _⟩;
+    rwa [(maximal_ideal K).eq_bot_of_prime, ideal.mem_bot] at *,
+    rw sub_self,
+  end }
+
+lemma henselian_local_ring.tfae (R : Type u) [comm_ring R] [local_ring R] :
+  tfae [
+  henselian_local_ring R,
+  ∀ (f : polynomial R) (hf : f.monic) (a₀ : residue_field R) (h₁ : aeval a₀ f = 0)
+    (h₂ : aeval a₀ f.derivative ≠ 0),
+    ∃ a : R, f.is_root a ∧ (residue R a = a₀),
+  ∀ {K : Type u} [field K], by exactI ∀ (φ : R →+* K) (hφ : surjective φ)
+    (f : polynomial R) (hf : f.monic) (a₀ : K) (h₁ : f.eval₂ φ a₀ = 0)
+    (h₂ : f.derivative.eval₂ φ a₀ ≠ 0),
+    ∃ a : R, f.is_root a ∧ (φ a = a₀)] :=
+begin
+  tfae_have _3_2 : 3 → 2, { intro H, exact H (residue R) ideal.quotient.mk_surjective, },
+  tfae_have _2_1 : 2 → 1,
+  { intros H, constructor, intros f hf a₀ h₁ h₂,
+    specialize H f hf (residue R a₀),
+    have aux := flip mem_nonunits_iff.mp h₂,
+    simp only [aeval_def, ring_hom.algebra_map_to_algebra, eval₂_at_apply,
+      ← ideal.quotient.eq_zero_iff_mem, ← local_ring.mem_maximal_ideal] at H h₁ aux,
+    obtain ⟨a, ha₁, ha₂⟩ := H h₁ aux,
+    refine ⟨a, ha₁, _⟩,
+    rw ← ideal.quotient.eq_zero_iff_mem,
+    rwa [← sub_eq_zero, ← ring_hom.map_sub] at ha₂, },
+  tfae_have _1_3 : 1 → 3,
+  { introsI hR K _K φ hφ f hf a₀ h₁ h₂,
+    obtain ⟨a₀, rfl⟩ := hφ a₀,
+    have H := henselian_local_ring.is_henselian f hf a₀,
+    simp only [← ker_eq_maximal_ideal φ hφ, eval₂_at_apply, φ.mem_ker] at H h₁ h₂,
+    obtain ⟨a, ha₁, ha₂⟩ := H h₁ _,
+    { refine ⟨a, ha₁, _⟩, rwa [φ.map_sub, sub_eq_zero] at ha₂, },
+    { contrapose! h₂,
+      rwa [← mem_nonunits_iff, ← local_ring.mem_maximal_ideal,
+        ← local_ring.ker_eq_maximal_ideal φ hφ, φ.mem_ker] at h₂, } },
+  tfae_finish,
+end
+
+instance (R : Type*) [comm_ring R] [hR : henselian_local_ring R] :
+  henselian_ring R (maximal_ideal R) :=
+{ jac := by { rw [ideal.jacobson, le_Inf_iff], rintro I ⟨-, hI⟩, exact (eq_maximal_ideal hI).ge },
+  is_henselian :=
+  begin
+    intros f hf a₀ h₁ h₂,
+    refine henselian_local_ring.is_henselian f hf a₀ h₁ _,
+    contrapose! h₂,
+    rw [← mem_nonunits_iff, ← local_ring.mem_maximal_ideal, ← ideal.quotient.eq_zero_iff_mem] at h₂,
+    rw h₂,
+    exact not_is_unit_zero
+  end }
+
+/-- A ring `R` that is `I`-adically complete is Henselian at `I`. -/
+@[priority 100] -- see Note [lower instance priority]
+instance is_adic_complete.henselian_ring
+  (R : Type*) [comm_ring R] (I : ideal R) [is_adic_complete I R] :
+  henselian_ring R I :=
+{ jac := is_adic_complete.le_jacobson_bot _,
+  is_henselian :=
+  begin
+    intros f hf a₀ h₁ h₂,
+    classical,
+    let f' := f.derivative,
+    -- in the following line, `f'.eval b` is a unit,
+    -- because `b` has the same residue class as `a₀`
```

Comment:
Loving the comments! Why not say what c is in words because regular humans don't know what `nat.rec_on` is? e.g.
```
    -- We now define a sequence `c n` by starting at `a₀` and then continually
    --  applying the function sending `b` to  `b - f(b)/f'(b)`. Note that `f'.eval b` is
    -- a unit, because `b` has the same residue as `a₀` modulo `I`.

Reply:
Thanks again!